### PR TITLE
Wip/emnlp autoqa camera ready

### PR DESCRIPTION
**Pull request by rayslxu**

### CODE REVIEW

File: tool/autoqa/schemaorg/manual-annotations.js [hunk 1]
```diff
@@ -183,6 +183,10 @@ const PROPERTY_CANONICAL_OVERRIDE = {
         base: ['location', 'address'],
         preposition: ["in #", "from #", "around #", "at #", "on #"]
     },
+    'postalCode': {
+        base: ['postal code', 'postcode', 'zip code'],
+        preposition: ['in #', 'from #'],
```
Comment (gcampax) [inline tool/autoqa/schemaorg/manual-annotations.js:194]: `in the # zip code` perhaps?

File: tool/autoqa/lib/canonical-extractor.js [hunk 2]
```diff
@@ -27,10 +27,13 @@ const child_process = require('child_process');
...
-    "2": "two",
-    "3": "three",
-    "4": "four"
+    "1": ["one"],
+    "2": ["two"],
+    "3": ["three"],
+    "4": ["four"],
+    "feb 14 2017": ["february 14 2017", "february 14, 2017", "feb 14, 2017", "14 february 2017", "14 february, 2017", "14 feb 2017", "14 feb, 2017"],
+    "may 4th, 2016": ["may 4 2016", "may 4, 2016", "may 4th 2016", "4th may 2016", "4th may, 2016", "4 may 2016", "4 may, 2016"],
+    "august 2nd 2017": ["august 2, 2017", "august 2 2017", "august 2nd, 2017", "2 august 2017", "2nd august 2017", "2 august, 2017", "2nd august, 2017"]
```
Comment (gcampax) [inline tool/autoqa/lib/canonical-extractor.js:37]: Is it worth using the tokenizer to extract these? We can already identify dates pretty well.

File: tool/autoqa/schemaorg/process-schema.js [hunk 3]
```diff
@@ -318,6 +318,8 @@ class SchemaProcessor {
                 metadata.counted_object = MANUAL_COUNTED_OBJECT_OVERRIDE[propertyname];
             else if (propertyname.startsWith('numberOf'))
                 metadata.counted_object = [ clean(propertyname.slice('numberOf'.length)) ];
+            else if (propertyname.startsWith('num') && propertyname.charAt(3) === propertyname.charAt(3).toUpperCase())
```
Comment (gcampax) [inline tool/autoqa/schemaorg/process-schema.js:321]: Writing this as a regular expression would be a lot readable...

File: tool/autoqa/schemaorg/manual-annotations.js [hunk 4]
```diff
@@ -351,6 +351,10 @@ const MANUAL_PROPERTY_CANONICAL_OVERRIDE = {
         verb_projection: ['know', 'understand', 'master'],
         adjective: ['# speaking']
     },
+    addressLocality: {
+        base: ['city'],
```
Comment (gcampax) [inline tool/autoqa/schemaorg/manual-annotations.js:355]: city, town, area, etc.

File: tool/autoqa/lib/bert-canonical-annotator.py [hunk 5]
```diff
@@ -212,6 +213,10 @@ def predict_one(self, table, arg, query, word, k):
         :param k: number of top candidates to return, this defaults to self.k if absent
         :return: a array in length k of predicted tokens
         """
+        # skip uncommon word
+        if not wordnet.synsets(word):
```
Comment (gcampax) [inline tool/autoqa/lib/bert-canonical-annotator.py]: Is it worth to bring in all of nltk and the wordnet dataset just to learn that a word is common or not? Seems like an overkill...

File: tool/autoqa/auto-annotate.js [hunk 6]
```diff
@@ -68,19 +68,14 @@ module.exports = {
...
+            help: 'Remove all existing canonical annotations in the schema',
             default: false
         });
-        parser.add_argument('--algorithm', {
-            help: 'Different algorithms to generate canonicals including bert, bart, adj, domain_synonyms split by comma (no space)',
-            default: null
+        parser.add_argument('--algorithms', {
+            nargs: '*',
+            help: 'Different algorithms to generate canonicals including bert-property-synonyms, bart-paraphrase, bert-adjectives, bert-domain-synonyms'
         });
```
Comment (gcampax) [inline tool/autoqa/auto-annotate.js:79]: What's the default here? Should we have a reasonable default? Is that BART only?

File: starter/schemaorg/Makefile [hunk 7]
```diff
@@ -91,13 +99,13 @@ custom_train_nlu_flags ?=
...
 .SECONDARY:
 
-models/paraphraser-bart:
-	mkdir models
-	wget --no-verbose https://almond-static.stanford.edu/test-data/paraphraser-bart.tar.xz
-	tar -C models -xvf paraphraser-bart.tar.xz
+models/paraphraser-bart-large-speedup-megabatch-5m:
+	mkdir -p models
+	wget --no-verbose https://almond-static.stanford.edu/test-data/paraphraser-bart-large-speedup-megabatch-5m.tar.xz
+	tar -C models -xvf paraphraser-bart-large-speedup-megabatch-5m.tar.xz
```
Comment (gcampax) [inline starter/schemaorg/Makefile:105]: Do we need to distribute both paraphraser models? Can we kill the old one?

**Review by gcampax — COMMENTED** [review]
Most of the changes are in autoqa's canonical generation, so I only gave a cursory look, cause I'm not very familiar with the code. If you want I can read more carefully, but I trust your tests really. I left a few comments here and there.

File: tool/autoqa/lib/canonical-extractor.js (hunk 2 above)
Comment (rayslxu) [reply tool/autoqa/lib/canonical-extractor.js:37]: Yeah, the extractor will be updated soon with an actual parser. I will do a separate PR to clean this up altogether.

File: tool/autoqa/lib/bert-canonical-annotator.py (hunk 5 above)
Comment (rayslxu) [reply tool/autoqa/lib/bert-canonical-annotator.py]: I searched around and didn't see a better solution... Any suggestion?

Comment (gcampax) [reply tool/autoqa/lib/bert-canonical-annotator.py]: We make our own vocabulary file (perhaps with wordnet as the starting point) and just load that. A single vocabulary file will be a lot smaller than nltk + wordnet.

File: tool/autoqa/auto-annotate.js (hunk 6 above)
Comment (rayslxu) [reply tool/autoqa/auto-annotate.js:79]: The current default is nothing - we skip auto annotation, but yeah, I think we should set it to the one that we think is the best

File: starter/schemaorg/Makefile (hunk 7 above)
Comment (rayslxu) [reply starter/schemaorg/Makefile:105]: you can kill the other one

Comment (gcampax) [reply starter/schemaorg/Makefile:105]: What i mean is overwrite the old model and keep the old URL, but point to the new model.

**Comment by rayslxu** [comment]
@gcampax I addressed most of the comments. 
For default options of algorithms, we will need a path to the paraphraser model if we want to set bart as the default, so I leave baseline as the default (skip auto annotate)
For the paraphraser model, I would rather keep the name.

**Pull request merged**